test/js-api/js-string/basic.tentative.any.js

```diff
@@ -0,0 +1,322 @@
+      results: [wasmRefType(kWasmExternRef)],
+    },
+    {
+      name: "substring",
+      params: [kWasmExternRef, kWasmI32, kWasmI32],
+      results: [wasmRefType(kWasmExternRef)],
+    },
+    {
+      name: "equals",
+      params: [kWasmExternRef, kWasmExternRef],
+      results: [kWasmI32],
+    },
+    {
+      name: "compare",
+      params: [kWasmExternRef, kWasmExternRef],
+      results: [kWasmI32],
+    },
+  ];
+
+  // Add a function type for each builtin
+  for (let builtin of builtins) {
+    builtin.type = builder.addType({
+      params: builtin.params,
+      results: builtin.results
+    });
+  }
+
+  // Add an import for each builtin
+  for (let builtin of builtins) {
+    builtin.importFuncIndex = builder.addImport(
+      "wasm:js-string",
+      builtin.name,
+      builtin.type);
+  }
+
+  // Generate an exported function to call the builtin
+  for (let builtin of builtins) {
+    let func = builder.addFunction(builtin.name + "Imp", builtin.type);
+    func.addLocals(builtin.params.length);
+    let body = [];
+    for (let i = 0; i < builtin.params.length; i++) {
+      body.push(kExprLocalGet);
+      body.push(...wasmSignedLeb(i));
+    }
+    body.push(kExprCallFunction);
+    body.push(...wasmSignedLeb(builtin.importFuncIndex));
+    func.addBody(body);
+    func.exportAs(builtin.name);
+  }
+
+  const buffer = builder.toBuffer();
+
+  // Instantiate this module using the builtins from the host
+  const builtinModule = new WebAssembly.Module(buffer, {
+    builtins: ["js-string"]
+  });
+  const builtinInstance = new WebAssembly.Instance(builtinModule, {});
+  builtinExports = builtinInstance.exports;
+
+  // Instantiate this module using the polyfill module
+  const polyfillModule = new WebAssembly.Module(buffer);
+  const polyfillInstance = new WebAssembly.Instance(polyfillModule, {
+    "wasm:js-string": polyfillImports
+  });
+  polyfillExports = polyfillInstance.exports;
+});
+
+// A helper function to assert that the behavior of two functions are the
+// same.
+function assert_same_behavior(funcA, funcB, ...params) {
+  let resultA;
+  let errA = null;
+  try {
+    resultA = funcA(...params);
+  } catch (err) {
+    errA = err;
+  }
+
+  let resultB;
+  let errB = null;
+  try {
+    resultB = funcB(...params);
+  } catch (err) {
+    errB = err;
+  }
+
+  if (errA || errB) {
+    assert_equals(errA === null, errB === null, errA ? errA.message : errB.message);
+    assert_equals(Object.getPrototypeOf(errA), Object.getPrototypeOf(errB));
+  }
+  assert_equals(resultA, resultB);
+
+  if (errA) {
+    throw errA;
+  }
+  return resultA;
+}
+
+// Constant values used in the tests below
+const testStrings = [
+  "",
+  "a",
+  "1",
+  "ab",
+  "hello, world",
+  "\n",
+  "☺",
+  "☺☺",
+  String.fromCodePoint(0x10000, 0x10001)
+];
+const testCharCodes = [1, 2, 3, 10, 0x7f, 0xff, 0xfffe, 0xffff];
+const testCodePoints = [1, 2, 3, 10, 0x7f, 0xff, 0xfffe, 0xffff, 0x10000, 0x10001];
+const testExternRefValues = [
```

Comment:
Is there a reason you don't test the exceptional cases here against the whole API?

Reply:
No good reason, just didn't think to do that. I'll see if I can work that in.